Content.Server/Radio/EntitySystems/MessagesServerSystem.cs

```diff
@@ -0,0 +1,129 @@
+            {
+                server.NextUpdate += server.UpdateDelay;
+
+                Update(uid, server);
+            }
+        }
+    }
+
+    public void Update(EntityUid uid, MessagesServerComponent component)
+    {
+        if (!this.IsPowered(uid, EntityManager))
+            return;
+
+        var serverQuery = EntityManager.AllEntityQueryEnumerator<MessagesServerComponent>();
+
+        while (serverQuery.MoveNext(out var serverUid, out var serverComponent))
+        {
+            component.Messages = new List<MessagesMessageData>(component.Messages.Union(serverComponent.Messages));
+            serverComponent.Messages = new List<MessagesMessageData>(component.Messages);
+        }
+
+        var packet = new NetworkPayload()
+        {
+            ["NameQuery"] = true,
+            ["ServerComponent"] = component
+        };
+        _deviceNetworkSystem.QueuePacket(uid, null, packet);
+    }
+
+    /// <summary>
+    /// Reacts to packets received from clients
+    /// </summary>
+    private void OnPacketReceived(EntityUid uid, MessagesServerComponent component, DeviceNetworkPacketEvent args)
+    {
+        if (!_singletonServerSystem.IsActiveServer(uid))
+            return;
+        if (args.Data.TryGetValue<string>("NewName", out var name) && args.Data.TryGetValue<int>("UserId", out var userId))
+            component.NameDict[userId] = name;
+        if (args.Data.TryGetValue<MessagesMessageData>("Message", out var message))
+            SendMessage(uid, component, message);
+    }
+
+    /// <summary>
+    /// Broadcasts a message into the network
+    /// </summary>
+    public void SendMessage(EntityUid uid, MessagesServerComponent component, MessagesMessageData message)
+    {
+        component.Messages.Add(message);
+
+        var packet = new NetworkPayload()
+        {
+            ["Message"] = message,
+            ["ServerComponent"] = component
+        };
+
+        _deviceNetworkSystem.QueuePacket(uid, null, packet);
+    }
+
+    /// <summary>
+    /// Returns the name of a given user
+    /// </summary>
+    public string GetNameFromDict(MessagesServerComponent component, int key)
+    {
+        if (component.NameDict.TryGetValue(key, out var value))
+            return value;
+        return Loc.GetString("messages-pda-user-missing");
+    }
+
+    /// <summary>
+    /// Returns the name dictionary cache
+    /// </summary>
+    public Dictionary<int, string> GetNameDict(MessagesServerComponent component)
+    {
+        return component.NameDict;
+    }
+
+    /// <summary>
+    /// Returns list of messages between the two users
+    /// </summary>
+    public List<MessagesMessageData> GetMessages(MessagesServerComponent component, int id1, int id2)
+    {
+        return new List<MessagesMessageData>(component.Messages.Where(message => message.SenderId == id1 && message.ReceiverId == id2 || message.SenderId == id2 && message.ReceiverId == id1));
+    }
+
```

Comment:
This is going to be extremely slow to the point you're better off just using 2 dictionaries of senders and receivers on top of the message list.